D1673/P1673.bs

```diff
@@ -5088,61 +5064,77 @@ satisfy the following requirements.
 All of these requirements presume that the algorithm's asymptotic complexity requirements,
 if any, are satisfied.
 
-1. Any input or output `mdspan`'s `value_type` meets the requirements of `semiregular`.
-
-2. The algorithm may perform read-only access on any input or output `mdspan`
-    arbitrarily many times.
-
-3. The algorithm or method may make arbitrary instances or copies
-    of any input or output `mdspan`'s `value_type`.
-
-4. The algorithm may assign arbitrarily many times to any reference
-    resulting from a valid output `mdspan` access,
-    as long as the algorithm's asymptotic complexity requirements are satisfied.
-
-5. The algorithm may assign any addend
-    in the algorithm's mathematical expression
-    to any reference resulting from a valid output `mdspan` access.
-
-6. The algorithm may reorder addends in its mathematical expression arbitrarily.
+1. Any input or output `mdspan`'s `value_type`
+    meets the requirements of `semiregular`.
+
+2. The algorithm may perform read-only access
+    on any input or output `mdspan` arbitrarily many times.
+
+3. The algorithm or method may make arbitrary many objects
+    of any input or output `mdspan`'s `value_type`,
+    value-initializing them or direct-initializing them
+    with any existing object of that type.
+
+4. The algorithm may assign arbitrarily many times
+    to any reference
+    resulting from a valid output `mdspan` access.
+
+5. If the algorithm's mathematical expression
+    uses addition,
+    but not subtraction, multiplication, or division,
+    then the algorithm evaluates the mathematical expression
+    using a sequence of evaluations of
+    `+`, `+=`, and `=` operators
+    that would produce the correct result
+    when operating on elements of a commutative semigroup.
+
+6. If the algorithm's mathematical expression
+    uses multiplication and possibly also addition,
+    but not subtraction or division,
+    then the algorithm
+    evaluates the mathematical expression
+    using a sequence of evaluations of
+    `*`, `*=`, `+`, `+=`, and `=` operators
+    that would produce the correct result
+    when operating on elements of a semiring
+    with noncommutative multiplication.
+
+7. If the algorithm's mathematical expression
+    uses subtraction and possibly also
+    addition and multiplication, but not division,
+    then the algorithm
+    evaluates the mathematical expression
+    using a sequence of evaluations of
+    `*`, `*=`, `+`, `+=`,
+    unary `-`, binary `-`, `-=`,
+    and `=` operators
+    that would produce the correct result
+    when operating on elements of a ring
+    with noncommutative multiplication.
+
+8. If the algorithm's mathematical expression
+    uses division and possibly also
+    addition, subtraction, and multiplication,
```

Comment:
You could reverse these to simplify the descriptions:

> 1. If the algorithm's mathematical expression _E_ uses division, then the algorithm evaluates _E_ using…
> 1. Otherwise, if _E_ uses subtraction, then the algorithm evaluates _E_ using…
> 1. Otherwise, if _E_ uses multiplication, …

Reply:
Ah, that's an excellent suggestion; thank you!  I might also add that we interpret "a / b" as "a times the inverse of b", to avoid confusion in the noncommutative case.

Reply:
I might actually want to do the opposite here: "If the algorithm's mathematical expression does _not_ use subtraction, then...."  I'm as much trying to exclude non-textbook algorithm choices (like Strassen's algorithm, for types that aren't all floating-point types), as I'm trying to show all permitted operations.  What do you think?

Reply:
My suggestion preserves meaning; yours does only if every expression uses a set of operations that is a prefix of `+*-/`.

Reply:
Fixed in the latest draft; thanks for the suggestion!